File: lesson_3_20240903/Sum.java
package lesson_3_20240903;

public class Sum {
    public static void main(String[] args) {
        int[] array = {1,2,3,4,5,6,7};
        System.out.println("Sum = " +sum(array));
        System.out.println("SumIteration = " +sumIteration(5));
        System.out.println("SumIteration = " +sumIterationReverse(5));
        System.out.println("SumRecursive = " +sumRecursive(5));
    }

    static int sum(int[] arr) {
        int sum = 0;
        for (int i = 0; i < arr.length; i++) {
            sum += arr[i];
        }
        return sum; 
    }

    static int sumIteration(int n) {
        int sum = 0;
        for (int i = 1; i <= n; i++) {
            sum += i;
        }
        return sum; 

    }

    static int sumIterationReverse(int n) {
        int sum = 0;
        for (int i = n; i > 0 ; i--) {
            sum += i;
        }
        return sum; 
    }

    
    static int sumRecursive(int n) {
        if(n==1) {
            return 1;
        }
        int sum = n + sumRecursive(n-1);
        return sum;
    }
    
    // sumRecursive(5) 
    //   5 + sumRecursive(4) 
    //      4 + sumRecursive(3) 
    //         3 + sumRecursive(2)   
    //            2 + sumRecursive(1)
    //               return 1
    //            return 2 + 1 = 3 
    //         return 3 + 3 = 6
    //      return 4 + 6 = 10
    //   return 5 + 10; 
    //  sum = 15  
}

File: lesson_3_20240903/sum.js
function sumIteration(n) {
    let sum = 0;
    for (let i = 1; i <= n; i++) {
        sum += i;
    }
    return sum; 

}

function sumIterationReverse(n) {
    let sum = 0;
    for (let i = n; i > 0 ; i--) {
        sum += i;
    }
    return sum; 
}


function sumRecursive(n) {
    if(n===1) {
        return 1;
    }
    let res = sumRecursive(n-1);
    let sum = n + res;
    console.log(`sum = ${sum}; n = ${n}`); 
    return sum;
}

// Пример косвенной рекурсии sumRecursive1 -> sumRecursive2 -> sumRecursive1
function sumRecursive1(n) {
    if(n===1) {
        return 1;
    }
    let res = sumRecursive2(n-1);
    let sum = n + res;
    console.log(`sum = ${sum}; n = ${n}`); 
    return sum;
}


function sumRecursive2(n) {
    if(n===1) {
        return 1;
    }
    let res = sumRecursive1(n-1);
    let sum = n + res;
    console.log(`sum = ${sum}; n = ${n}`); 
    return sum;
}


console.log("Sum = "+sumIteration(5))
console.log("SumReverse = "+sumIterationReverse(5))
console.log("SumRecursive = "+sumRecursive(5))
console.log("SumRecursive1 = "+sumRecursive1(5))

    
    // sumRecursive(5) res = 15
    //   sumRecursive(4) res = 10
    //     sumRecursive(3) res = 6
    //         sumRecursive(2) res = 3  
    //            sumRecursive(1) res = 1
    //                  return 1
    //            2 + 1 = 3, return 3 (sumRecursive(2))
    //         3 + 3 = 6, return 6 (sumRecursive(3))
    //      4 + 6 = 10, return 10 (sumRecursive(4))
    //   5 + 10 = 15, return 15 (sumRecursive(5))
    //  sum = 15  